extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -0,0 +1,178 @@
+              LOGGER.debug("Query {} executed {} rows affected", sql, rowsUpdated);
+            } catch (SQLException e) {
+              LOGGER.error("Error executing query {}", sql, e);
+              // re:throw this as unhandled exception
+              throw new RuntimeException(e);
+            }
+            sqlBuffer.setLength(0); // Clear the buffer for the next statement
+          }
+        }
+      }
+    } catch (IOException e) {
+      LOGGER.error("Error reading the script file", e);
+      throw new RuntimeException(e);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing the script file", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  public <T> List<T> executeSelect(String query, Class<T> targetClass) {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement();
+        ResultSet s = statement.executeQuery(query)) {
+      List<T> results = ResultSetToObjectConverter.convert(s, targetClass);
+      return results.isEmpty() ? null : results;
+    } catch (Exception e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  public int executeUpdate(String query) {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement()) {
+      return statement.executeUpdate(query);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      return handleException(e);
+    }
+  }
+
+  public int executeUpdate(String query, Statement statement) {
+    LOGGER.debug("Executing query {} within transaction", query);
+    try {
+      return statement.executeUpdate(query);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      return handleException(e);
+    }
+  }
+
+  public void runWithinTransaction(TransactionCallback callback) {
+    Connection connection = null;
+    try {
+      connection = borrowConnection();
+      connection.setAutoCommit(false); // Disable auto-commit to start a transaction
+
+      boolean result;
+      try (Statement statement = connection.createStatement()) {
+        result = callback.execute(statement);
+      }
+
+      if (result) {
+        connection.commit(); // Commit the transaction if successful
+      } else {
+        connection.rollback(); // Rollback the transaction if not successful
+      }
+
+    } catch (SQLException e) {
+      if (connection != null) {
+        try {
+          connection.rollback(); // Rollback on exception
+        } catch (SQLException ex) {
+          LOGGER.error("Error rolling back transaction", ex);
+        }
+      }
+      LOGGER.error("Caught Error while executing transaction", e);
+      handleException(e);
+    } finally {
+      if (connection != null) {
+        try {
+          connection.setAutoCommit(true); // Restore auto-commit
+          connection.close();
```

Comment:
Could we use try-with-resources instead of closing the connection explicitly?

Reply:
`connection` is an `AutoCloseable`... Why not do `try(Connection connection = ... ) { come code; }`?

Reply:
but don't we need to rollback in case we encounter SQLException (Lets say constraint violation) ? for that on catching the sqlException ? so if I add this IDE complains this : 
![Screenshot 2025-04-16 at 1 41 15 PM](https://github.com/user-attachments/assets/44691f71-5df7-4495-af0a-e691d1d38ff3)

Reply:
restructured it to get rid of finally to have two try's